server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java

```diff
@@ -0,0 +1,288 @@
+                }
+            }
+
+            @Override
+            public void delete() {
+                ensureNotClosed();
+                final String thisPath = path.buildAsString();
+                synchronized (context.actions) {
+                    consistentView(context.actions).stream().filter(action -> action.path.startsWith(thisPath))
+                        .forEach(a -> context.actions.add(new BlobStoreAction(Operation.DELETE, a.path)));
+                }
+            }
+
+            @Override
+            public Map<String, BlobMetaData> listBlobs() {
+                ensureNotClosed();
+                final String thisPath = path.buildAsString();
+                synchronized (context.actions) {
+                    return consistentView(context.actions).stream()
+                        .filter(
+                            action -> action.path.startsWith(thisPath) && action.path.substring(thisPath.length()).indexOf('/') == -1
+                                && action.operation == Operation.PUT)
+                        .collect(
+                            Collectors.toMap(
+                                action -> action.path.substring(thisPath.length()),
+                                action -> new PlainBlobMetaData(action.path.substring(thisPath.length()), action.data.length)));
+                }
+            }
+
+            @Override
+            public Map<String, BlobContainer> children() {
+                ensureNotClosed();
+                final String thisPath = path.buildAsString();
+                synchronized (context.actions) {
+                    return consistentView(context.actions).stream()
+                        .filter(action ->
+                            action.operation == Operation.PUT
+                                && action.path.startsWith(thisPath) && action.path.substring(thisPath.length()).indexOf('/') != -1)
+                        .map(action -> action.path.substring(thisPath.length()).split("/")[0])
+                        .distinct()
+                        .collect(Collectors.toMap(Function.identity(), name -> new MockBlobContainer(path.add(name))));
+                }
+            }
+
+            @Override
+            public Map<String, BlobMetaData> listBlobsByPrefix(String blobNamePrefix) {
+                return Maps.ofEntries(
+                    listBlobs().entrySet().stream().filter(entry -> entry.getKey().startsWith(blobNamePrefix)).collect(Collectors.toList())
+                );
+            }
+
+            @Override
+            public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists)
+                throws IOException {
+                ensureNotClosed();
+                // TODO: Throw if we try to overwrite any blob other than incompatible_snapshots or index.latest with different content
```

Comment:
what about this TODO? Can we enable this already for some files?

Reply:
Yea we can for all but `snap-` (at the root level) and `index-N` at the root level I think. Will do :)

Reply:
I way beefed this up now. We were only running into this for `snap-` with the current implementation (since we don't have list inconsistencies simulated yet index-N isn't an issue). 
For the `snap-` I actually made it check for the case when whatever is written only differs in timstamps (on a master failover this can happen) and let it pass.
Shard level overwrites I made fail 100% of the time (see comment I added for that, it's not expected).
I think this is pretty neat now. This hasn't failed over 10k+ runs of `SnapshotResiliencyTests` and shows that we're handling master fail-overs reasonably safely on S3 (obviously not testing list inconsistencies yet ... but still :D).